AsyncDisplayKit/Layout/ASLayoutSpec.mm

```diff
@@ -113,28 +112,35 @@ - (void)setParent:(id<ASLayoutable>)parent
   }
 }
 
-- (void)setChild:(id<ASLayoutable>)child;
+- (void)setChild:(id<ASLayoutable>)child
 {
   ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
-  
+  ASDisplayNodeAssertNotNil(child, @"Child cannot be nil");
+
   id<ASLayoutable> finalLayoutable = [self layoutableToAddFromLayoutable:child];
-  _child = finalLayoutable;
-  [self propagateUpLayoutable:finalLayoutable];
+  if (finalLayoutable) {
+    _children = @[finalLayoutable];
+    [self propagateUpLayoutable:finalLayoutable];
+  }
 }
 
 - (void)setChild:(id<ASLayoutable>)child forIdentifier:(NSString *)identifier
 {
   ASDisplayNodeAssert(self.isMutable, @"Cannot set properties when layout spec is not mutable");
+  ASDisplayNodeAssertNotNil(child, @"Child cannot be nil");
   
   id<ASLayoutable> finalLayoutable = [self layoutableToAddFromLayoutable:child];
   self.childrenWithIdentifier[identifier] = finalLayoutable;
+  if (finalLayoutable) {
+    self.children = [self.children arrayByAddingObject:finalLayoutable];
```

Comment:
Hmm, I don't really understand this code.  If I set a child for an existing identifier, it should not be appended to the children array, right?  Do we even need both?

I think we should consider separating out these changes.  I'd love to reduce dependence on the dictionary in this class, as it shows up materially in profiles, but I'm a bit worried about some nonintuitive risks here that only rarer layout code exposes.

Anyway, once we've confirmed that unconditionally adding the object is correct, two quick things:
- If self.children is nil, it's not clear to me this will work - may need to alloc the array
- we should not use self.children and instead use _children, because self.children creates a copy, which we're about to copy again by adding an object.

This code needs to be locked in the near future, right?  Maybe add a FIXME somewhere.